crates/sage-cloudpath/src/tdf.rs

```diff
@@ -53,16 +158,186 @@ impl TdfReader {
+    // Make sure the mz array is sorted
+    // In theory I could use the type system to enforce this but I dont
+    // think it is worth it, its not that slow and its simple.
+    assert!(mz_array.windows(2).all(|x| x[0] <= x[1]), "mz_array is not sorted");
+
+    let arr_len = mz_array.len();
+    let mut global_num_included = 0;
+
+    let mut order: Vec<usize> = (0..arr_len).collect();
+    order.sort_unstable_by(|&a, &b| {
+        intensity_array[b]
+            .partial_cmp(&intensity_array[a])
+            .unwrap_or(Ordering::Equal)
+    });
+
+    #[derive(Clone, Copy)]
+    struct ImsPeak {
+        mz: f32,
+        intensity: f32,
+        im: f32,
+    }
+    let mut agg_buff = Vec::with_capacity(10_000.min(arr_len));
+    // TODO: Optimize the size of this buffer.
+    const INCLUDE_BUFF_SIZE: usize = 1000;
+    let mut included = vec![false; arr_len];
+    let mut include_buff = [false; INCLUDE_BUFF_SIZE];
+
+    let utol = mz_tol_ppm / 1e6;
+    let im_tol = im_tol_pct / 100.0;
+
+    for &idx in &order {
+        if included[idx] {
+            continue;
+        }
+
+        let mz = mz_array[idx];
+        let im = ims_array[idx];
+        let da_tol = mz * utol;
+        let left_e = mz - da_tol;
+        let right_e = mz + da_tol;
+
+        let mut ss_start = mz_array.partition_point(|&x| x < left_e);
+        let mut ss_end = mz_array.partition_point(|&x| x <= right_e);
+
+        let mut slice_width = ss_end - ss_start;
+        if slice_width >= 1000 {
+            // It is EXCEEDINGLY UNLIKELY that more than 1000 points
+            // will be aggregated or in the range of a single mz peak.
+            // Here we just handle those edge cases by making sure the 'center'
+            // will be aggregated.
+
+            let new_ss_start = idx.saturating_sub(INCLUDE_BUFF_SIZE / 2);
+            let new_ss_end = new_ss_start + INCLUDE_BUFF_SIZE - 1;
+            // TODO: make a better warning message here.
+            log::warn!(
+                "More than {} points are in the mz range of a point, limiting span",
+                INCLUDE_BUFF_SIZE
+            );
+            ss_start = new_ss_start;
+            ss_end = new_ss_end;
+            slice_width = ss_end - ss_start;
+        }
+
+        let abs_im_tol = im * im_tol;
+        let left_im = im - abs_im_tol;
+        let right_im = im + abs_im_tol;
+
+        let mut curr_intensity = 0.0;
+
+        let mut num_includable = 0;
+        for i in ss_start..ss_end {
+            let im_i = ims_array[i];
+            if !included[i] && intensity_array[i] > 0.0 && im_i >= left_im && im_i <= right_im {
+                curr_intensity += intensity_array[i];
+                num_includable += 1;
+                include_buff[i - ss_start] = true;
+            }
+        }
+
+        agg_buff.push(ImsPeak {
+            mz,
+            intensity: curr_intensity,
+            im,
+        });
+        included[ss_start..ss_end]
```

Comment:
Note: There is a bug here ... I am including only the points that match the mz+ims but flagging as included all points int the mz range ... should also add an 'if' in the `for_each` to flag only what actually got included (which is the whole reason I have a buffer ...)

Reply:
Fixed in: [6e52396](https://github.com/lazear/sage/pull/166/commits/6e5239613b0ec0f096ea395c0e80be6d345f11da)